msyslog-v1.08a+smac/src/man/im_tcp.8
.\"	$CoreSDI: im_tcp.8,v 1.3.2.6 2001/11/20 09:56:23 alejo Exp $
.\"
.\" Copyright (c) 2001
.\"	Core-SDI SA. All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of Core-SDI SA nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.Dd Feb 28, 2001
.Dt IM_TCP 8
.Os Core-SDI
.Sh NAME
.Nm tcp input module
.Nd
.Xr syslogd 8
input module used to receive from tcp connections
.Sh SYNOPSIS
.Nm tcp
.Op Fl a
.Op Fl q
.Op Fl h Ar host
.Op Fl p Ar port
.Sh DESCRIPTION
.Nm tcp input module 
allows
.Xr syslogd 8
daemon to use tcp IPC.
.Sh EXAMPLES
.Bl -bullet
.It
The following command installs
.Xr syslogd 8
daemon and activates tcp system accepting remote connections on local
address machinename.example.com port 3210:
.Pp
.Dl syslogd -i 'tcp -h machinename.somedomainexample.com -p 3210'
.El
.Bl -bullet
.It
The following command installs
.Xr syslogd 8
daemon and activates tcp system accepting remote connections on all local
addresses port 3210:
.Pp
.Dl syslogd -i 'tcp -p 3210'
.El
.Bl -bullet
.It
The following is the similar to the previous, but extracts the host name
from the message:
.Pp
.Dl syslogd -i 'tcp -a -p 3210'
.It
The following is the same, but avoids adding full fomain (FQDN):
.Pp
.Dl syslogd -i 'tcp -q -p 3210'
.Pp
.Sh BUGS
.Bl -bullet
.It
We are sending everything not encrypted! You may want to use a tunnel
such as SSL.
.It
Submit bugs at this project's Sourceforge Bug reporting system at:
http://sourceforge.net/tracker/?func=add&group_id=25741&atid=385117
You may also report them directly to the authors; send an email to
core.devel.alat@corest.com, describing the problem the most you can,
containing also machine description, hardware description, the
configuration file (/etc/syslog.conf), the OS description, and the
invoking command line.
The more you describe the bug, the faster we can fix it.
.El
.Sh SEE ALSO
.Xr syslog 3 ,
.Xr im_bsd 8 ,
.Xr im_doors 8 ,
.Xr im_linux 8 ,
.Xr im_streams 8 ,
.Xr im_unix 8 ,
.Xr im_udp 8 ,
.Xr om_tcp 8 ,
.Xr syslogd 8


html/ChangeLog.txt
$Id$

Feature Request:

Bugs:
1.  Change code to allow for tailing spaces when looking at the expression field

3/10/2004
	-  Removing any reference to company name

3/9/2004
	-  Changed out company logo
	-  Added GPL notice to all libraries

1/21/2004
	-  Found a bug where a host would email off an alert with no text.  I've put a stop gap fix in.

1/13/2004
        -  Fixed bug with msyslog.pgsql where SET STORAGE didn't have a space before it.
        -  Fixed bug where hour in time of rules would default to 18:00 for no good reason
	-  Fixed two problems with cloning rules:  1)  order was not preserved  2)  an imported version of the database couldn't clone because fields that were empty needed to have non-null defaults applied
	-  Forgot to merge in latest VPN reports

1/6/2004
        -  Added 'left menu' support for ACID and MRTG groups in security framework should the ever be added 8)

1/6/2004
	-  Added 'left menu' support for ACID and MRTG groups in security framework should the ever be added 8)

1/1/2004
	-  Fixed spelling error in with the word 'threshold' in the rules schema
	-  Fixed a bug in vacuumdb where it was 'ANALYZ' not 'ANALYZE'

12/19/2003
	-  Updated code to look into /opt/apache

12/10/2003
	-  Updated database to schema to not use compressed text fields(we'll see how this performs)
	-  working on adding interfaces to more lock data and other new stats with PostgreSQL V7.4
	-  cleaned up some button descriptions on the maintenance page, also added a lock view as well as a settings view

12/9/2003
	-  Finished rule.php support for basic timer maintenance.  Need to add another page to graft timers onto rules.
	-  processlogs.php is now setup to support rule timers
	-  need to write rule expiration process 
	-  started converting away from compressed text in hopes of providing faster data retrieval.... also pulled OIDs from DB definition

12/8/2003
	-  Adding support for date and time based rules with date ranges, day of week selections, as well as deleted rules, need to add interface to control rule timestamp properties

12/7/2003
	-  finished adding accumulation thresholds

12/3/2003
	-  adding support for both types of thresholds.  Need to update web pages to reflect new radio buttons.
	-  added web configuration support for supression thresholds and accumulating thresholds, now onto updating the log processor
	-  basic supression works(kinda), need to verify functionality

12/2/2003
	-  customer profiles can now have multiple hosts added at once.
	-  fixed a stupid bug where <prev|refresh|next> didn't behave right.  next would stop working if you started at oct-26-2003 and it would stay on oct-26-2003
	-  Added web-based framework and database schema to support alert supression thresholds

12/1/2003
	-  Updated processlogs.php code to better deal with single entry/no rules vs mutliple rules

11/19/2003
	-  Updated mail table to enforce unique login ids(effectively stopping two processes from running at the same time, one will crash and die(safely))
	-  Updated openmail and closeopenmail to use transaction support since PostgreSQL no longer does server-side auto-commit(ie. convert everything to transactions)
	-  Transaction support should now be officially added, will do some testing

11/18/2003
	-  Updating program for support with PostGreSQL V7.4
	-  Fixed host process table to reflect the fact that multiple hosts are in there by default	

10/13/2003
	-  Added alert total to bottom of alert aggregation

10/6/2003
	-  finished support for alert aggregation
	-  updated 1stview to pull the current time and date
	-  fixed host.php bug where you could expire syslogs but not be forced to expire syslogs

10/5/2003
	-  still adding support for alert aggregation.  Basic aggregation works plus alert zooming but need to add support for across the board for other alert queries

10/3/2003
	-  added ability to un/suspend log processors from the web interface
	-  updated maintenance to rebuild all indexes in an better manner(ie. grab the index list from the DB rather then by hand)
	-  added additional framework to do alert aggregation interface.  Need to add 'aggregation code' for display

9/24/2003
	-  weeklyindexrebuild.php now pulls all indexes from the system and rebuilds them.  The result is that the system will now rebuild any new indexes without manual reconfiguration.

9/15/2003
	-  launchid was not initiated correctly in the clonerule.

9/2/2003
	-  Fixed a bug where '\'s at the end of a line caused problems because we were not properly dealing with them in general.  Fixed that.

8/23/2003	
	-  Created another bug when fixing 5000 line paging.  Timestamp was thrown off in view.php

8/18/2003	
	-  Syncing changes from production smt environment:  vacuumtsyslog.php
	-  Updated weeklyindexrebuild.php to account for the correct indexes
	-  Updated maint.php to account for the three new indexes for the launch program section

8/13/2003
	-  view.php has had several updates.  Paging should now be fixed.  Multiple searches appeared not to be working correctly.
	-  Needed to add lastid as hidden var if the variable was set
	-  Needed to use urlencode on top of htmlspecialcharacters, filters were broken because of it

8/11/2003
	-  1stfilter.php doesn't list 'global' filters that you down own
	-  modified view.php to not let the user save a filter with no description
	-  added support to delete all of a user's filters(ie. do before delete)
	-  another problem popped up with filter.php when I added the delete user filter option

8/1/2003	
	-  processlogs.php now supports launching external programs!

7/31/2003
	-  Updated vacuumdb.php to do a full vacuum of the TSyslog table.  Why?  Because the system doesn't reclaim disk space or use old delete space for some reason
	-  Almost finished adding launch program code, need to test.

7/27/2003
	-  Added weeklyindexrebuild.php which rebuilds all indexes at 5am Sunday morning

7/23/2003

msyslog-v1.08a+smac/src/modules/im_tcp.c
/*	$CoreSDI: im_tcp.c,v 1.16.2.6 2001/11/30 23:13:44 alejo Exp $	*/

/*
 * Copyright (c) 2001, Core SDI S.A., Argentina
 * All rights reserved
 * 
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither name of the Core SDI S.A. nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 * IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

/*
 * im_tcp -- input from INET using TCP
 *
 * Author: Alejo Sanchez for Core SDI S.A.
 * 
 * This input module is a bit tricky, because of the nature of TCP
 * connections, and the use of poll() for I/O on syslogd
 * 
 * The main idea is that first a im_tcp module will be called
 * and it will bind to a port and wait for connections to it.
 * 
 * Whenever a conection is established it will add it to an
 * array of file descriptors of connections.
 * 
 */

#include "config.h"


#if TIME_WITH_SYS_TIME
# include <sys/time.h>
# include <time.h>
#else
# if HAVE_SYS_TIME_H
#  include <sys/time.h>
# else
#  include <time.h>
# endif
#endif
#include <sys/types.h>
#include <sys/socket.h>
#include <sys/uio.h>
#include <ctype.h>
#include <stdlib.h>
#include <unistd.h>
#include <errno.h>
#include <signal.h>
#include <string.h>
#include <stdio.h>
#include <syslog.h>

#include "../modules.h"
#include "../syslogd.h"

/* recvfrom() and others like socklen_t, Irix doesn't provide it */   
#ifndef HAVE_SOCKLEN_T
  typedef int socklen_t;
#endif

struct tcp_conn {
	struct tcp_conn *next;
	int		 fd;
	char		 name[MAXHOSTNAMELEN + 1];
	char		 port[20];
	char		 saveline[MAXLINE + 3]; /* maxline + cr lf */
};

struct im_tcp_ctx {
	socklen_t	 addrlen;
	struct tcp_conn	*first;
	struct tcp_conn	*last;
	int		flags;
};

#define	M_USEMSGHOST	0x01
#define	M_NOTFQDN	0x02


void printline(char *, char *, size_t, int);
int listen_tcp(char *host, char *port, socklen_t *);
int accept_tcp(int, socklen_t, char *, int, char *, int);


/*
 * initialize tcp input
 *
 * this module takes a host argument (ie. 0.0.0.0, 0::0, server.example.com)
 * and a port/service ('syslog' or numerical)
 *
 */

int
im_tcp_init(struct i_module *I, char **argv, int argc)
{
	struct im_tcp_ctx	*c;
	char			*host, *port;
	int			ch, optind_s;

	dprintf(MSYSLOG_INFORMATIVE, "im_tcp_init: entering\n");

	if ( (I->im_ctx = calloc(1, sizeof(struct im_tcp_ctx))) == NULL) {
		dprintf(MSYSLOG_SERIOUS, "om_tcp_init: cant alloc memory");
		return (-1);
	}

	c = (struct im_tcp_ctx *) I->im_ctx;

	host = "0.0.0.0";
	port = "syslog";

	/* parse command line */
	optind_s = optind;	/* save main's optind */
	optind = 1;
#ifdef HAVE_OPTRESET
	optreset = 1;
#endif
	while ((ch = getopt(argc, argv, "h:p:aq")) != -1) {
		switch (ch) {
		case 'h':
			/* get addr to bind */
			host = optarg;
			break;
		case 'p':
			/* get remote host port */
			port = optarg;
			break;
		case 'a':
			c->flags |= M_USEMSGHOST;
			break;
		case 'q':
			/* don't use domain in hostname (FQDN) */
			c->flags |= M_NOTFQDN;
			break;
		default:
			dprintf(MSYSLOG_SERIOUS, "om_tcp_init: parsing error"
			    " [%c]\n", ch);
			free(c);
			return (-1);
		}
	}

	optind = optind_s;	/* restore main's optind */

	if ( (I->im_fd = listen_tcp(host, port, &c->addrlen)) < 0) {
		dprintf(MSYSLOG_SERIOUS, "im_tcp_init: error with listen_tcp() %s\n",
		    strerror(errno));
		free(c);
		return (-1);
	}

	I->im_path = NULL;

	add_fd_input(I->im_fd , I);

	dprintf(MSYSLOG_INFORMATIVE, "im_tcp_init: running\n");

	return (1);
}


/*
 * im_tcp_read: accept a connection and add it to the queue
 *
 */

int
im_tcp_read(struct i_module *im, int infd, struct im_msg *ret)
{
	struct im_tcp_ctx *c;
	struct tcp_conn *con;
	int n;

	dprintf(MSYSLOG_INFORMATIVE, "im_tcp_read: entering...\n");

	if (im == NULL || ret == NULL) {
		dprintf(MSYSLOG_SERIOUS, "im_tcp_read: arg %s%s is null\n",
		    ret? "ret":"", im? "im" : "");
		return (-1);
	}

	if ((c = (struct im_tcp_ctx *) im->im_ctx) == NULL) {
		dprintf(MSYSLOG_SERIOUS, "im_tcp_read: null context\n");
		return (-1);
	}

	if (infd == im->im_fd) {

		dprintf(MSYSLOG_INFORMATIVE, "im_tcp_read: new connection\n");

		/* create a new connection */
		if ((con = (struct tcp_conn *) calloc(1, sizeof(*con)))
		    == NULL) {
			dprintf(MSYSLOG_SERIOUS, "im_tcp_read: "
			    "error allocating conn struct\n");
			return (-1);
		}

		/* accept it and add to queue */
		if ((con->fd = accept_tcp(infd, c->addrlen, con->name,
		    sizeof(con->name), con->port, sizeof(con->port))) < 0) {
			dprintf(MSYSLOG_SERIOUS, "im_tcp_read: couldn't accept\n");
			free (con);
			return (-1);
		}

		/* add to queue */
		if (c->last == NULL) {
			c->first = con;
		} else {
			c->last->next = con;
		}
		c->last = con;


		dprintf(MSYSLOG_INFORMATIVE, "im_tcp_read: new conection from"
		    " %s with fd %d\n", con->name, con->fd);

		/* add to inputs list */
		add_fd_input(con->fd , im);

		return (0); /* 0 because there is no line to log */

	}

	/* read connected socket */

	dprintf(MSYSLOG_INFORMATIVE, "im_tcp_read: reading connection fd %d\n",
	    infd);

	/* find connection */
	for (con = c->first; con && con->fd != infd; con = con->next);

	if (con == NULL || con->fd != infd) {
		dprintf(MSYSLOG_SERIOUS, "im_tcp_read: no such connection "
		    "fd %d !\n", infd);
		remove_fd_input(infd);
		return (-1);
	}

	n = read(con->fd, im->im_buf, sizeof(im->im_buf) - 1);
	if (n == 0) {
		struct tcp_conn *prev;

		dprintf(MSYSLOG_INFORMATIVE, "im_tcp_read: conetion from %s"
		    " closed\n", con->name);

		remove_fd_input(con->fd);

		/* connection closed, remove its tcp_con struct */
		close (con->fd);

		if (con == c->first) {
			c->first = con->next;
			if (con == c->last)
				c->last = NULL;
		} else {
			for(prev = c->first; prev->next != con;
			    prev = prev->next);
			prev->next = con->next;
		}

		if (con->saveline[0] != '\0')
			printline(ret->im_host, con->saveline,
			    strlen(con->saveline),  0);

		free(con);
	
		return (0);

	} else if (n < 0 && errno != EINTR) {
		dprintf(MSYSLOG_INFORMATIVE, "im_tcp_read: conetion from %s"
		    " closed with error [%s]\n", con->name, strerror(errno));
		logerror("im_tcp_read");
		con->fd = -1;
		remove_fd_input(con->fd);
		return (0);
	} else {
		char	*p, *nextline, *cr;

		/* terminate it */
		(im->im_buf)[n] = '\0';
		p = &im->im_buf[0];

		dprintf(MSYSLOG_INFORMATIVE, "im_tcp_read: read: %s [%s]",
		    con->name, im->im_buf);

		/* change non printable chars to X, just in case */
		for(p = im->im_buf; *p != '\0'; p++)
			if (!isprint((unsigned int) *p) && *p != '\n')
				*p = 'X';
		p = im->im_buf;

		do {
			char	*msg;

			msg = p;

			/* multiple lines ? */
			if((nextline = strchr(p, '\n')) != NULL) {
				/* terminate this line and advance */
				*nextline++ = '\0';
				if (*nextline == '\0')
					nextline = NULL; /* no more lines */
			} else {
				/* save this partial line and return */
				strncat(con->saveline, p,
				    sizeof(con->saveline) - 1);
				con->saveline[sizeof(con->saveline) - 1] = '\0';
			}

			/* remove trailing carriage returns */
			if ((cr = strchr(p, '\r')) != NULL)
				*cr = '\0';

			if (*p == '\0') {
				if (nextline != NULL) {
					p = nextline;
					continue;
				} else
					return (0);
			}

			if (c->flags & M_USEMSGHOST) {
				char	host[90];
				int	n1, n2;

				if (con->saveline[0] != '\0') {
					strncat(con->saveline, p,
					    sizeof(con->saveline) - 1);
					con->saveline[sizeof(con->saveline) - 1]
					    = '\0';
					msg = con->saveline;
				} else {
					msg = p;
				}

				/* extract hostname from message */
				if ((sscanf(msg, "<%*d>%*3s %*i %*i:%*i:%*i %n%89s"
				    " %n", &n1, host, &n2) != 1 &&
				    sscanf(msg, "%*3s %*i %*i:%*i:%*i %n%89s %n",
				    &n1, host, &n2) != 1 &&
				    sscanf(msg, "%n%89s %n", &n1,
				    host, &n2) != 1)
				    || im->im_buf[n2] == '\0') {
					dprintf(MSYSLOG_INFORMATIVE,
					    "im_tcp_read: ignoring invalid "
					    "message [%s]\n", msg);
					if (nextline != NULL) {
						p = nextline;
						continue;
					} else {
						return (0);
						con->saveline[0] = '\0';
					}
				}

				/* remove host from message */
				while (im->im_buf[n2] != '\0')
					im->im_buf[n1++] = im->im_buf[n2++];
				im->im_buf[n1] = '\0';

				strncpy(ret->im_host, host,
				    sizeof(ret->im_host) - 1);
				ret->im_host[sizeof(ret->im_host) - 1] = '\0';

			} else {

				/* get hostname from originating addr */
				strncpy(ret->im_host, con->name,
				    sizeof(ret->im_host) - 1);
				ret->im_host[sizeof(ret->im_host) - 1] = '\0';
			}

			if (c->flags & M_NOTFQDN) {
				char	*dot;

				if ((dot = strchr(ret->im_host, '.')) != NULL)
					*dot = '\0';
			}

			printline(ret->im_host, msg,  strlen(msg),  0);
			*msg = '\0';

			p = nextline;

		} while (nextline != NULL);
	}

	return (0); /* we already logged the lines */
}

int
im_tcp_close(struct i_module *im)
{
	struct im_tcp_ctx *c;
	struct tcp_conn *con, *cnext;

	c = (struct im_tcp_ctx *) im->im_ctx;

	/* close all connections */
	for (con = c->first; con; con = cnext) {
		if (con->fd > -1)
			close(con->fd);
		cnext = con->next;
		free(con);
	}

	im->im_ctx = NULL;

	/* close listening socket */
	return (close(im->im_fd));
}
